docs/usage.md

```diff
@@ -1,76 +1,80 @@
+during the build itself, requiring that all dependencies are declared and pre-fetched before the build occurs.
+
+In order to support this class of hermetic builds, not only does Cachi2 need to pre-fetch the dependencies, but some
+build flows will need additional changes (i.e. leveraging defined [environment variables](#generate-environment-variables)
+or using Cachi2 to [inject project files](#inject-project-files)).
 
 ### Pre-fetch dependencies
 
+The first step in creating hermetic builds is to fetch the dependencies for one of the [supported package managers](../README.md#package-managers).
+
 ```shell
 cachi2 fetch-deps \
-  --source ./fzf \
+  --source ./foo \
   --output ./cachi2-output \
-  '{"path": ".", "type": "gomod"}'
+  '{"path": ".", "type": "<supported package manager>"}'
 ```
 
 * `--source` - the path to a *git repository* on the local disk
 * `--output` - the path to the directory where Cachi2 will write all output
 * `{JSON}`   - specifies a *package* (a directory) within the repository to process
 
 Note that Cachi2 does not auto-detect which package managers your project uses. You need to tell Cachi2 what to process
-when calling fetch-deps. In the example above, the package is a go module located at the root of the fzf repo,
+when calling fetch-deps. In the example above, the package path is located at the root of the foo repo,
 hence the relative path is `.`.
 
 The main parameter (PKG) can handle different types of definitions:
 
-* simple: `gomod`, same as `{"path": ".", "type": "gomod"}`
-* JSON object: `{"path": "subpath/to/other/module", "type": "gomod"}`
-* JSON array: `[{"path": ".", "type": "gomod"}, {"path": "subpath/to/other/module", "type": "gomod"}]`
+* simple: `<package manager>`, same as `{"path": ".", "type": "<package manager>"}`
+* JSON object: `{"path": "subpath/to/other/module", "type": "<package manager>"}`
+* JSON array: `[{"path": ".", "type": "<package manager>"}, {"path": "subpath/to/other/module", "type": "<package manager>"}]`
 * JSON object with flags:
-`{"packages": [{"path": ".", "type": "gomod"}], "flags": ["gomod-vendor"]}`
+`{"packages": [{"path": ".", "type": "<package manager>"}], "flags": ["gomod-vendor"]}`
 
 See also `cachi2 fetch-deps --help`.
 
+Using the JSON array object, multiple package managers can be used to resolve dependencies in the same repository.
+
+*⚠ While Cachi2 does not intentionally modify the source repository unless the output and source paths are the same,
+some package managers may add missing data like checksums as dependency data is resolved. If this occurs from a clean
+git tree then the tree has the possibility to become dirty.*
+
 ### Generate environment variables
 
-```shell
-cachi2 generate-env ./cachi2-output -o ./cachi2.env --for-output-dir /tmp/cachi2-output
-```
+Once the dependencies have been cached, the build process needs to be made aware of the dependencies. Some package
```

Comment:
Funny thing is that Cachi2 will probably only ever care about the prefetching. :stuck_out_tongue:

Reply:
@chmeliik

Reply:
Caching is relevant still as we are caching locally/on some volume to be made available at build time so that they do not have to be re-retrieved.

Some forms of caching do not need to be handled differently from the package manager's perspective (i.e. leveraging squid proxies for all network traffic). Other forms need to be integrated with a package manager to be able to be leveraged. If you wanted to, I am sure that you could come up with a much more obtuse cache-awareness method to teach a package manager.

Even using Nexus requires to teach the package manager as you have to direct the manager to the location.

> prefetch: identify and download content that's needed to perform a network-isolated build

I wouldn't include the latter part. This is just about a uniform process to identify and download content before building. Network isolation is out of scope for pre-fetch as the onus for being able to run in these environments is completely on the user to provide a complete set. Pre-fetching _can_ be used for other things too like speeding up builds if dependencies are pre-fetched asynchronously -- especially if network speeds are slow.
> cache: a layer that speeds up the prefetching of dependencies

You are describing a pre-fetch cache here. Cachito has this as it knows the references for the pre-fetched content. Cachi2 doesn't (I assume) as it just bases the pre-fetch off of the local content. I think a more relevant type of cache is to speed up/enable builds, again in network-restrictive situations. Cachi2 _can_ be used in contexts outside of container images as well -- i.e. if you want to be able to run python packages in network-restricted environments.
> permanent storing: everything that is built once can be rebuilt again ~without needing to access the Internet~

Maybe without needing to access the original sources instead of the internet? Ensuring that you have adequate caches available that are saved off certainly helps with this part of reproducibility.

Reply:
If we consistently use prefetch, we can avoid philosophical questions :slightly_smiling_face: 

> Cachito has this as it knows the references for the pre-fetched content. Cachi2 doesn't (I assume) as it just bases the pre-fetch off of the local content.

Cachi2 doesn't because we haven't implemented one (and maybe we never will). There is no real difference in what Cachito/cachi2 knows, cachi2 just doesn't have Athens/Nexus to store stuff in.

Reply:
If you don't try to address the philosophical questions, you won't know how to relate to the outside world! :D 

A _generic_ cache (like a squid proxy) is _some_ form of a pre-fetch cache as it brings content closer to the cluster and would speed up repeated retrievals. Another form of pre-fetch caches would be if _entire_ prefetches were cached (i.e. to S3). The former wouldn't be implemented within Cachi2, the latter would likely have to be to some extent.